ADPhantomApp/src/ADPhantom.cpp

```diff
@@ -3167,6 +3297,33 @@ asynStatus ADPhantom::readoutDataStream(int start_cine, int end_cine, int start_
     }
     //for (int frame = 0; frame < frames; frame++){
     while ((frame < frames) && (status == asynSuccess)){
+      getIntegerParam(PHANTOM_DownloadAbort_, &abort);
+      if(abort){
+        setStringParam(ADStatusMessage, "Download aborting");
+        //To abort cleanly we disconnect from the port to restart the datastream
+        debug(functionName, "Running common connect");
+        status = pasynCommonSyncIO->connect(dataPort_, 0, &commonDataport_, NULL);
+        if (status){
+          debug(functionName, "Common connect failed");
+        }
+        debug(functionName, "Running pasynCommonSyncIO->disconnectDevice");
+        status = pasynCommonSyncIO->disconnectDevice(commonDataport_);
+        if (status) {
+          debug(functionName, "Disconnect device failed");
+        }
+        debug(functionName, "Running pasynCommonSyncIO->connectDevice");
+        status = pasynCommonSyncIO->connectDevice(commonDataport_);
+        if (status) {
+          debug(functionName, "Connect device failed");
+        }
+        debug(functionName, "Running attachToPort");
+        status = attachToPort("dataPort");
+        if(status){
+          debug(functionName, "Failed to attach ");
+        }
+        break;
+      }
+
       metaFrame = start_frame+frame;
       frame++;
       total_frame++;
```

Comment:
Interesting that the call to readFrame maintains the lock.  If the application is responsive whilst downloading cines with lots of frames then I think this is OK.